src/analytical_solutions/sedov_solution/main.cpp

```diff
@@ -92,36 +85,47 @@ int main(int argc, char** argv)
     vector<double> Px (nParts);
     vector<double> Py (nParts);
     vector<double> Pz (nParts);
+    vector<ParticleIO> vSim(nParts);
 
-    // Load particles data
-    FileData::readData3D(
-        inputFile,
-        nParts,
-        x,y,z,
-        vx,vy,vz,
-        h,
-        rho,u,p,
-        cs,
-        Px,Py,Pz);
-
-    // Calculate radius, velocity and sort particle data by radius
-    vector<Particle> vSim(nParts);
-    for(size_t i = 0; i < nParts; i++)
+    if (!only_sol)
     {
-        double r   = sqrt( pow(x[i], 2) + pow(y[i], 2) + pow(z[i], 2) );
-        double vel = sqrt( pow(vx[i],2) + pow(vy[i],2) + pow(vz[i],2) );
-
-        vSim[i] = {i,
-                   r,vel,
-                   x[i],y[i],z[i],
-                   vx[i],vy[i],vz[i],
-                   h[i],
-                   rho[i],u[i],p[i],
-                   cs[i],
-                   Px[i],Py[i],Pz[i]};
-    }
-    sort(vSim.begin(),vSim.end(),Particle::cmp());
+        if (nParts <= 0){
+            cout << "ERROR: --nParts: '" << nParts << "' should be > 0." << endl;
+            exit(EXIT_FAILURE);
+        }
+        else if (!filesystem::exists(inputFile)){
+            cout << "ERROR: --input file: '" << inputFile << "' don't exist." << endl;
+            exit(EXIT_FAILURE);
+        }
 
+        // Load particles data
+        FileData::readData3D(
+            inputFile,
+            nParts,
+            x,y,z,
+            vx,vy,vz,
+            h,
+            rho,u,p,
+            cs,
+            Px,Py,Pz);
+
+        // Calculate radius, velocity and sort particle data by radius
+        for(size_t i = 0; i < nParts; i++)
+        {
+            double r   = sqrt( pow(x[i], 2) + pow(y[i], 2) + pow(z[i], 2) );
+            double vel = sqrt( pow(vx[i],2) + pow(vy[i],2) + pow(vz[i],2) );
+
+            vSim[i] = {i,
+                       r,vel,
+                       x[i],y[i],z[i],
+                       vx[i],vy[i],vz[i],
+                       h[i],
+                       rho[i],u[i],p[i],
+                       cs[i],
+                       Px[i],Py[i],Pz[i]};
+        }
+        sort(vSim.begin(),vSim.end(),ParticleIO::cmp());
```

Comment:
Side remark: sorting particle data according to some specified criterion is a standard problem in SPH and has a standard solution. We're doing this on every SPH step to sort the particles according to ascending space filling curve key. The approach taken here has the downside that you first have to define a custom data layout `struct ParticleIO` and copy the original data there before it can be sorted. You have to pay the memory space for a full duplication of all the particles data and also, if the number of particle fields change, you have to adapt your data layout.

A standard way of solving this problem involves using a function called `sort_by_key`, it sorts an array of "values" according to another array of provided "keys". In your example, the keys are the radii, and the values are an `std::iota` sequence (0, 1, 2, 3, 4, 5, ...). The effect is that the iota sequence will be changed into the ordering that sorted your keys. You can then apply this ordering to as many particle buffers as you like.

An example (`test/integration_mpi/domain_nranks.cpp`):
```c++
    computeSfcKeys(x.data(), y.data(), z.data(), sfcKindPointer(codesGlobal.data()), numParticles, box);
    std::vector<LocalIndex> ordering(numParticles);
    std::iota(begin(ordering), end(ordering), LocalIndex(0));
    sort_by_key(begin(codesGlobal), end(codesGlobal), begin(ordering));
    reorderInPlace(ordering, x.data());
    reorderInPlace(ordering, y.data());
    reorderInPlace(ordering, z.data());
    reorderInPlace(ordering, h.data());
```

`sort_by_key` is for instance supported by the `Thrust` library, which means that we can immediately also apply this approach to buffers on the GPU. In addition, we don't need a custom sort key and the sorting happens on elementary types, which means that the implementation of `sort_by_key` can use O(N) radix-sort instead of O(NlogN) merge sort.

Reply:
The ParticleData is now integrated. I am checking how to sort.

Reply:
Reordering in place